src/call/buffers.rs

```diff
@@ -0,0 +1,570 @@
+        if self.remaining == 0 {
+            return Ok(0);
+        }
+        buf.reserve(self.remaining);
+        let start = buf.len();
+        let mut len = start;
+        unsafe {
+            buf.set_len(start + self.remaining);
+        }
+        let ret = loop {
+            match self.read(&mut buf[len..]) {
+                Ok(0) => break Ok(len - start),
+                Ok(n) => len += n,
+                Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+                Err(e) => break Err(e),
+            }
+        };
+        unsafe {
+            buf.set_len(len);
+        }
+        ret
+    }
+}
+
+impl BufRead for MessageReader {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        // Optimization for empty slice
+        if self.remaining == 0 {
+            return Ok(&[]);
+        }
+
+        // When finished reading current `buffer_slice`, start reading next slice
+        let buffer_len = self.buffer_slice.len();
+        if buffer_len == 0 || self.buffer_offset == buffer_len {
+            self.buffer_slice = self.reader.next_slice();
+            self.buffer_offset = 0;
+        }
+
+        debug_assert!(self.buffer_offset <= buffer_len);
+        Ok(self.buffer_slice.range_from(self.buffer_offset))
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.remaining -= amt;
+        self.buffer_offset += amt;
+    }
+}
+
+#[cfg(feature = "prost-codec")]
+impl Buf for MessageReader {
+    fn remaining(&self) -> usize {
+        self.remaining
+    }
+
+    fn bytes(&self) -> &[u8] {
+        // This is similar but not identical to `BuffRead::fill_buf`, since `self`
+        // is not mutable, we can only return bytes up to the end of the current
+        // slice.
+
+        // Optimization for empty slice
+        if self.buffer_slice.len() == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.buffer_offset <= self.buffer_slice.len());
+        self.buffer_slice.range_from(self.buffer_offset)
+    }
+
+    fn advance(&mut self, mut cnt: usize) {
+        // Similar but not identical to `BufRead::consume`. We must also advance
+        // the buffer slice if we have exhausted the current slice.
+
+        // The number of bytes remaining in the current slice.
+        let mut remaining = self.buffer_slice.len() - self.buffer_offset;
+        while remaining <= cnt {
+            self.consume(remaining);
+            if self.remaining == 0 {
+                return;
+            }
+
+            cnt -= remaining;
+            self.buffer_slice = self.reader.next_slice();
+            self.buffer_offset = 0;
+            remaining = self.buffer_slice.len();
+        }
+
+        self.consume(cnt);
+    }
+}
+
+/// A zero-copy writer.
+///
+/// This is implemented by writing into a Rust `Vec`, then wrapping that with the
+/// necessary gRPC data structures (see `vec_slice`).
+pub struct MessageWriter {
+    pub write_buffer: Vec<u8>,
```

Comment:
Why not use `grpc_slice` directly?

Reply:
Because it is slow! I got a very decent speedup by using a `Vec` over any of the grpc primitives. I looked into why, but I don't recall the details. I think `grpc_slice` did a lot more allocation.